include/deal.II/fe/fe_update_flags.h

```diff
@@ -336,6 +346,187 @@ namespace CellSimilarity
+
+      /**
+       * Array of the inverse Jacobian matrices at the quadrature points.
+       */
+      std::vector<DerivativeForm<1,spacedim,dim> > inverse_jacobians;
+
+      /**
+       * Array of quadrature points. This array is set up upon calling reinit()
+       * and contains the quadrature points on the real element, rather than on
+       * the reference element.
+       */
+      std::vector<Point<spacedim> >  quadrature_points;
+
+      /**
+       * List of outward normal vectors at the quadrature points. This field is
+       * filled in by the finite element class.
+       */
+      std::vector<Point<spacedim> >  normal_vectors;
+
+      /**
+       * List of boundary forms at the quadrature points. This field is filled in
+       * by the finite element class.
+       */
+      std::vector<Tensor<1,spacedim> >  boundary_forms;
+    };
+
+
+    /**
+     * A class that stores all of the shape function related data used in
+     * dealii::FEValues, dealii::FEFaceValues, and dealii::FESubfaceValues
+     * objects. Objects of this kind will be given
+     * as <i>output</i> argument when dealii::FEValues::reinit()
+     * calls FiniteElement::fill_fe_values().
+     *
+     * @ingroup feaccess
+     */
+    template <int dim, int spacedim=dim>
+    class FiniteElementRelatedData
+    {
+    public:
+      /**
+       * Initialize all vectors to correct size.
+       */
+      void initialize (const unsigned int        n_quadrature_points,
+                       const FiniteElement<dim,spacedim> &fe,
+                       const UpdateFlags         flags);
+
+      /**
+       * Storage type for shape values. Each row in the matrix denotes the values
+       * of a single shape function at the different points, columns are for a
+       * single point with the different shape functions.
+       *
+       * If a shape function has more than one non-zero component (in deal.II
+       * diction: it is non-primitive), then we allocate one row per non-zero
+       * component, and shift subsequent rows backward.  Lookup of the correct row
+       * for a shape function is thus simple in case the entire finite element is
+       * primitive (i.e. all shape functions are primitive), since then the shape
+       * function number equals the row number. Otherwise, use the
+       * #shape_function_to_row_table array to get at the first row that belongs
+       * to this particular shape function, and navigate among all the rows for
+       * this shape function using the FiniteElement::get_nonzero_components()
+       * function which tells us which components are non-zero and thus have a row
+       * in the array presently under discussion.
+       */
+      typedef dealii::Table<2,double> ShapeVector;
+
+      /**
+       * Storage type for gradients. The layout of data is the same as for the
+       * #ShapeVector data type.
+       */
+      typedef std::vector<std::vector<Tensor<1,spacedim> > > GradientVector;
+
```

Comment:
I like the patch in general, but have one comment about the type of GradientVector and HessianVector: Now that you're touching this data anyway, why not use dealii::Table<2,Tensor<1,spacedim> > straight away instead of the clumsy `vector<vector<Tensor>>` data structure? This reduces the memory allocations and makes code easier. I'm not sure how much work it would be to adjust this in the various Mapping::transform methods which use VectorSlice<vector> arguments, though.